otherlibs/unix/fork.c

```diff
@@ -36,10 +36,12 @@ CAMLprim value unix_fork(value unit)
   if (ret == 0) caml_atfork_hook();
   if (ret == -1) uerror("fork", Nothing);
 
-  CAML_EVENTLOG_DO({
-      if (ret == 0)
-        caml_eventlog_disable();
-  });
+  if (ret == 0) {
+    caml_eventring_post_fork();
+    CAML_EV_LIFECYCLE(EV_FORK_CHILD, 0);
+  } else {
+    CAML_EV_LIFECYCLE(EV_FORK_PARENT, ret);
+  }
```

Comment:
I see. Would the intention be that multiple parts of the runtime could register hooks for atfork? Is there an example elsewhere where we're doing something similar?

Reply:
I think the first step would be to have well-identified functions that "do whatever needs to be done" right after a fork, one for the child and one for the parent. Otherwise all naked calls to `fork()` in the compiler codebase have to duplicate logic, and of course we will see bugs.

A second step would be, possibly, to make it possibly to register hooks indeed, instead of having all io-state-maintainers within the runtime dit otherlibs/unix/fork.c with their own stuff. This could come as a later PR for example.

In the C runtime there are many types of hooks (just grep for `hook` there), usually it's just one function pointer (often atomic), and the user is expected to (1) save the previous hook if non-NULL somewhere, (2) set their own hook, (3) ensure that their own hooks calls the (saved) previous hook. This gives some composability, but I think it's rather fragile -- I'm not sure it actually works if several people set and then unset their hooks in non-bracketed order.

For example:

https://github.com/ocaml/ocaml/blob/094725c1b5fe8e118a2b9635dda18de1c4f1e5ff/otherlibs/systhreads/st_stubs.c#L131-L162

On the OCaml side, there are some hooks like `at_exit` for which the stdlib explicitly maintains a queue, removing opportunities for users to write non-compositional code.

https://github.com/ocaml/ocaml/blob/094725c1b5fe8e118a2b9635dda18de1c4f1e5ff/stdlib/stdlib.ml#L557-L568

Reply:
I see, thanks. That makes sense - I'll get that logic separated out on Monday.

Reply:
> A second step would be, possibly, to make it possibly to register hooks indeed, instead of having all io-state-maintainers within the runtime dit otherlibs/unix/fork.c with their own stuff. This could come as a later PR for example.

POSIX threads have `pthread_atfork`, which does exactly what you're after.  And it looks like we agreed to use the POSIX threads API even on native Windows ports (#11115).  So let's not reinvent another hook registration mechanism.

This said, I'm not a big fan of hooks and would be happy with just well-identified functions to call before fork / after fork in the parent / after fork in the child.

Reply:
Is the change in https://github.com/ocaml/ocaml/pull/10964/commits/fd2da1246e5079139e76a0e6b740bb9ac4d15071 what you had in mind @gasche ?